src/librustc_middle/ty/query/mod.rs

```diff
@@ -189,3 +189,23 @@ pub fn force_from_dep_node<'tcx>(tcx: TyCtxt<'tcx>, dep_node: &DepNode) -> bool
 pub(crate) fn try_load_from_on_disk_cache<'tcx>(tcx: TyCtxt<'tcx>, dep_node: &DepNode) {
     rustc_dep_node_try_load_from_on_disk_cache!(dep_node, tcx)
 }
+
+/// An analogue of the `Into` trait that's intended only for query paramaters.
+///
+/// This exists to allow queries to accept either `DefId` or `LocalDefId` while requiring that the
+/// user call `to_def_id` to convert between them everywhere else.
+pub trait IntoQueryParam<P> {
+    fn into_query_param(self) -> P;
+}
```

Comment:
In case there was any confusion, "sealing" works like this:
```rust
mod sealed {
    pub trait IntoQueryParam<P> {
        fn into_query_param(self) -> P;
    }
}
```

Then you can refer to it as `sealed::IntoQueryParam`, and as long as `sealed` remains private, and you don't reexport `IntoQueryParam` somewhere, it should be impossible to implement or import the trait, outside of the parent module of the `mod sealed`.

Reply:
Ah, I thought you meant the super-trait patten that's in the API guidelines.

Reply:
@ecstatic-morse ah I see, that pattern uses a sealed trait for a specific purpose.
Or maybe I'm the only person referring to "`pub` but in a private `mod`" as "sealed".